lib/fs/basicfs_watch.go

```diff
@@ -38,10 +39,16 @@ func (f *BasicFilesystem) Watch(name string, ignore Matcher, ctx context.Context
 	}
 
 	absShouldIgnore := func(absPath string) bool {
+		if !utf8.ValidString(absPath) {
```

Comment:
We also need this check on L93/100. I am not 100% sure, but I think even if notify is working perfectly it's not guaranteed that the filter is applied: It might only be applied to watches that are set up, i.e. parent directories - the event of a file inside that directory might still be invalid. And even if that's not the case, I wouldn't trust it fully.

Reply:
Maybe it's best to do all these in ignore.go instead then to have one place where files are filtered out? Could repurpose Match() to be a universal filter (check file type/name + ignore patterns, etc.) instead of sanitizing separately in several places. It can return a Result instead of a bool.

Reply:
> Maybe it's best to do all these in ignore.go instead then to have one place where files are filtered out? Could repurpose Match() to be a universal filter (check file type/name + ignore patterns, etc.) instead of sanitizing separately in several places. It can return a Result instead of a bool.

I can see how always checking in `ignore` is attractively simple/general/failure proof. And `Match()` already does return a result type (since a very short time), so that doesn't even need to change (just a new value to be added).

My gut feeling is still to designate `ignore` as requiring valid utf-8 input like most of the code, and filtering non-utf-8 strings in lib/fs when using `Matcher`. However that's a weakly held opinion based on the conflicting arguments I try to explain below - I am very ready/likely to be convinced otherwise. Or maybe to not protract this too much, just get a third opinion as a tie-breaker - as in of course feel free to not engage in this discussion further if you don't feel like it:

I see how it's magic/non-obvious that `lib/scanner` acts as an utf-8-validness-barrier for code sitting behind or in front of it. And the magic is that if "one knows", then it's clear that `lib/fs` and related code is in front, while everything else (famously descriptive "model", i.e. all of database, syncing, protocol) is behind it. At the same time it's also nice to be able to just postulate that everything is valid utf-8 and not need to think about it in that code.

I was initially hesitant performance wise as it will be redundant in most places, however I subsequently tried to make a good argument and couldn't. Respectively only for the scanner, and there we already do that check which we could now skip. Then again it's also not overly nice to do/hide utf-8 validation in `ignore.Matcher` - seems to violate the "least surprise" principle to me, and I like that one. Then again we drop internal files on `ignore.Match` - not entirely obvious from the name either.

Reply:
Honestly you know the code better than me, so I'll concede on it. My best argument is that it feels more correct to have all the validation and checks in one place so unexpected panics like this don't happen in the future because some part of the code forgot to do a check or the are at least easier to track down.

Reply:
> Honestly you know the code better than me [...]

That's a double edge sword - it also makes me subsceptible to conservatism/favouring the status quo :)